insta/src/runtime.rs

```diff
@@ -641,17 +697,41 @@ pub fn assert_snapshot(
 
     let tool_config = get_tool_config(manifest_dir);
 
-    // apply filters if they are available
-    #[cfg(feature = "filters")]
-    let new_snapshot_value =
-        Settings::with(|settings| settings.filters().apply_to(new_snapshot_value));
+    if let Some(snapshot_file) = &ctx.snapshot_file {
+        Snapshot::cleanup_extra_files(ctx.old_snapshot.as_ref(), snapshot_file)?;
+    }
+
+    let content = match new_snapshot_value {
+        SnapshotValue::Text { content, .. } | SnapshotValue::InlineText { content, .. } => {
+            // apply filters if they are available
+            #[cfg(feature = "filters")]
+            let content = Settings::with(|settings| settings.filters().apply_to(content));
 
-    let kind = match ctx.snapshot_file {
-        Some(_) => SnapshotKind::File,
-        None => SnapshotKind::Inline,
+            let kind = match ctx.snapshot_file {
+                Some(_) => TextSnapshotKind::File,
+                None => TextSnapshotKind::Inline,
+            };
+
+            TextSnapshotContents::new(content.into(), kind).into()
+        }
+        SnapshotValue::Binary {
+            content, extension, ..
+        } => {
+            assert!(extension != "new", "this file extension is not allowed");
```

Comment:
not important for now, but as we get closer, we should do things like include `new` in the error message here